certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java

```diff
@@ -0,0 +1,205 @@
+     *
+     * @param id is the {@link long} to find.
+     * @return {@link Optional<GiftCertificate>}.
+     */
+    @Override
+    public Optional<GiftCertificate> findById(long id) {
+        return entityManager
+                .createQuery("select c from Certificate c where c.id = :id", GiftCertificate.class)
+                .setParameter("id", id)
+                .getResultList().stream().findFirst();
+    }
+
+    /**
+     * Updates the {@link GiftCertificate}.
+     *
+     * @param entity is the value of the {@link GiftCertificate} to update.
+     */
+    @Override
+    public void update(GiftCertificate entity) {
+        GiftCertificate giftCertificate = entityManager.find(GiftCertificate.class, entity.getId());
+        giftCertificate.setName(entity.getName());
+        giftCertificate.setDescription(entity.getDescription());
+        giftCertificate.setPrice(entity.getPrice());
+        giftCertificate.setDuration(entity.getDuration());
+        giftCertificate.setTags(entity.getTags());
+    }
+
+    /**
+     * Deletes the {@link GiftCertificate} object from the database.
+     *
+     * @param id is the value of the {@link long} to find.
+     */
+    @Override
+    public void delete(long id) {
+        giftCertificateRepository.deleteById(id);
+    }
+
+    /**
+     * Finds {@link Optional<GiftCertificate>} in the database by the name of the {@link GiftCertificate}.
+     *
+     * @param name is the {@link String} to find.
+     * @return {@link Optional<GiftCertificate>}.
+     */
+    @Override
+    public Optional<GiftCertificate> findByName(String name) {
+        return entityManager
+                .createQuery("select c from Certificate c where c.name = :name", GiftCertificate.class)
+                .setParameter("name", name)
+                .getResultList().stream().findFirst();
+//        return giftCertificateRepository.findByName(name);
+    }
+
+    /**
+     * Saves certificateId and tagId in the database.
+     *
+     * @param certificateId is the id of the {@link GiftCertificate} to save.
+     * @param tagId         is the id of the {@link CertificateTag} to save
+     */
+    @Override
+    public void saveIdsInHas_tagTable(long certificateId, long tagId) {
+
+    }
+
+    /**
+     * Removes the tuple certificateId and tagId from the 'has_tag' table of the database.
+     *
+     * @param certificateId is the id of the {@link GiftCertificate} to remove.
+     * @param tagId         is the id of the {@link CertificateTag} to remove.     *
+     */
+    @Override
+    public void deleteIdsInHas_TagTable(long certificateId, Long tagId) {
+
+    }
+
+    /**
+     * Finds certificate without {@link CertificateTag} by its name.
+     *
+     * @param name the name to find by.
+     * @return {@link Optional<GiftCertificate>}.
+     */
+    @Override
+    public Optional<GiftCertificate> findCertificateWithoutTagsByName(String name) {
+        return Optional.empty();
+    }
```

Comment:
Why do you keep such empty methods? What do you know about SOLID (exactly **I** letter)?

Reply:
Because i use this methods in the jdbcDao. And they are the part of the interface and should be overriden in the new realisation.
Yes, interface segregation should be done. Just now i have no idea how to implement it.

Reply:
done